modules/programs/borgmatic.nix

```diff
@@ -122,12 +134,24 @@ let
 
   removeNullValues = attrSet: filterAttrs (key: value: value != null) attrSet;
 
+  hmFiles = lib.mapAttrsToList (_: file: file) config.home.file;
+  hmSymlinks = (lib.filter (file: !file.recursive) hmFiles);
+  hmExcludePattern = file: "${config.home.homeDirectory}/${file.target}";
+  hmExcludePatterns = lib.concatMapStringsSep "\n" hmExcludePattern hmSymlinks;
+  hmExcludeFile = pkgs.writeText "hm-symlinks.txt" hmExcludePatterns;
+  hmSymlinksExclusions = config:
+    lib.optionalAttrs config.location.excludeHomeManagerSymlinks {
+      exclude_from =
+        (lib.optionals (config.location.extraConfig ? "exclude_from")
+          config.location.extraConfig.exclude_from) ++ [ hmExcludeFile ];
```

Comment:
The ad hoc merging here could be avoided by moving this whole thing into the `configModule` submodule.

Reply:
I [tried to do that](https://gist.github.com/DamienCassou/8b35027f35ad8f81b9657cf78d55c859) but it fails because `location.extraConfig.exclude_from` can be used in user configuration files as well:

```
error: The option `programs.borgmatic.backups.main.location.extraConfig.exclude_from' has conflicting definition values:
       - In `/home/cassou/Documents/projects/nix-system/home-manager/modules/programs/borgmatic.nix':
           [
             <derivation /nix/store/h010kdss4pd2lb0j2zl4fmnl57k4p2yg-hm-symlinks.txt.drv>
           ]
       - In `/home/cassou/Documents/projects/nix-system/home-manager/tests/modules/programs/borgmatic/exclude-hm-symlinks.nix':
           [
             <derivation /nix/store/a7az4qq2b9w5cknk6jnhacxm9nizwzyx-excludeFile.txt.drv>
           ]
```

Reply:
I feel the solution could be to make `exclude_from` a declared option instead of an extra one. 

What do you think?

Reply:
I think that error is because the type of `extraConfig` is `attrsOf (oneOf [ str bool path int (listOf str) ])`, but you're trying to merge two lists of *derivations*, which aren't strings, so it doesn't know how to merge them. `listOf (either str path)` should fix it.

Reply:
Well, since this is YAML we might as well just use the YAML type from [`pkgs.formats`](https://nixos.org/manual/nixos/unstable/#sec-settings-nix-representable).

Reply:
I [tried your suggestion](https://gist.github.com/DamienCassou/8b35027f35ad8f81b9657cf78d55c859) but it still shows the same error message. Can we merge this PR and try that in a separate one?

Reply:
Applying that patch on top of the current revision I get no eval error, only a test error that is fixed by swapping 0 and 1 in the index to `exclude_from`. We might want to make the order more deterministic using `mkBefore` or `mkAfter`.

Reply:
Ok, I applied the patch and the CI is happy so I am happy.

For some reason, the test still fails locally for me when using `nix-shell`:

```
$ nix-shell --pure tests -A run.borgmatic-program-exclude-hm-symlinks
error: The option `programs.borgmatic.backups.main.location.extraConfig.exclude_from' has conflicting definition values:
       - In `/home/cassou/Documents/projects/nix-system/home-manager/tests/modules/programs/borgmatic/exclude-hm-symlinks.nix':
           [
             "/nix/store/7vvpsb5d589890x125agz60d5i7jsxav-excludeFile.txt"
           ]
       - In `/home/cassou/Documents/projects/nix-system/home-manager/modules/programs/borgmatic.nix':
           [
             "/nix/store/ix4zmh6b02az1lwm98q1wzsn2rbarlsz-hm-symlinks.txt"
           ]
(use '--show-trace' to show detailed location information)
```

The problem also appears when doing `home-manage switch`.

But everything is fine even locally when using `nix develop`:

```
$ nix develop --ignore-environment .#tests.borgmatic-program-exclude-hm-symlinks
borgmatic-program-exclude-hm-symlinks: OK
```

Could there be a difference between nixos-unstable and release-22.11?

I'm happy merging the PR like it is though.